test_runner/regress/test_compatibility.py

```diff
@@ -0,0 +1,181 @@
+import os
+import re
+import shutil
+import subprocess
+from pathlib import Path
+from typing import Any
+
+import pytest
+import toml
+from fixtures.neon_fixtures import (
+    NeonCli,
+    NeonEnvBuilder,
+    NeonPageserverHttpClient,
+    PgBin,
+    wait_for_last_record_lsn,
+    wait_for_upload,
+)
+from fixtures.types import Lsn
+from pytest import FixtureRequest
+
+
+def dump_differs(first: Path, second: Path, output: Path) -> bool:
+    """
+    Runs diff(1) command on two SQL dumps and write the output to the given output file.
+    Returns True if the dumps differ, False otherwise.
+    """
+
+    with output.open("w") as stdout:
+        rv = subprocess.run(
+            [
+                "diff",
+                "--unified",  # Make diff output more readable
+                "--ignore-matching-lines=^--",  # Ignore changes in comments
+                "--ignore-blank-lines",
+                str(first),
+                str(second),
+            ],
+            stdout=stdout,
+        )
+
+    return rv.returncode != 0
+
+
+def test_backward_compatibility(pg_bin: PgBin, test_output_dir: Path, request: FixtureRequest):
+    compatibility_snapshot_dir_env = os.environ.get("COMPATIBILITY_SNAPSHOT_DIR")
+    assert (
+        compatibility_snapshot_dir_env is not None
+    ), "COMPATIBILITY_SNAPSHOT_DIR is not set. It should be set to `compatibility_snapshot_pg14` path generateted by test_prepare_snapshot"
+    compatibility_snapshot_dir: Path = Path(compatibility_snapshot_dir_env).resolve()
+
+    # Make compatibility snapshot artifacts pickupable by Allure
+    # by copying the snapshot directory to the curent test output directory.
+    repo_dir = test_output_dir / "compatibility_snapshot" / "repo"
+
+    shutil.copytree(compatibility_snapshot_dir / "repo", repo_dir)
+
+    # Remove old logs to avoid confusion in test artifacts
+    for logfile in repo_dir.glob("**/*.log"):
+        logfile.unlink()
+
+    # Update paths in configs
+    pageserver_config = (repo_dir / "pageserver.toml").read_text()
+    pageserver_config = pageserver_config.replace(
+        "/test_prepare_snapshot/", "/test_backward_compatibility/compatibility_snapshot/"
+    )
+    (repo_dir / "pageserver.toml").write_text(pageserver_config)
+
+    for postmaster_opts in repo_dir.glob("**/postmaster.opts"):
+        postmaster_opts_content = postmaster_opts.read_text()
+        postmaster_opts_content = postmaster_opts_content.replace(
+            "/test_prepare_snapshot/", "/test_backward_compatibility/compatibility_snapshot/"
+        )
+        postmaster_opts.write_text(postmaster_opts_content)
+
+    snapshot_config = toml.load(repo_dir / "config")
+
+    # NeonEnv stub to make NeonCli happy
```

Comment:
Were there some roadblocks when you tried to use default env? In my understanding using env manually without calling to init should be fine. WDYT?

Reply:
Also since real env is a context manager you dont have to manually add finalizers. Env will properly shutdown itself

Reply:
Do you mean `NeonEnvBuilder` / `neon_env_builder` fixture?
That's a nice idea, I've tried to use it, but `start` doesn't start without `init`, unfortunately:
https://github.com/neondatabase/neon/blob/f5e9d3010bfb6d0fc5cfa2aaba971b668ae9d69a/test_runner/fixtures/neon_fixtures.py#L556-L563
And cooking up env manually (or rather from a config) will require some refactoring. I'd like to do it, but probably in a separate PR and later.

Reply:
At a quick glance doesnt look too wordy, but up to you